### PR TITLE
TF-TRT Pack op conversion in dynamic shape mode

### PR DESCRIPTION
**Pull request by tfeher**
This PR improves the Pack (Stack) op converter to handle explicit batch and dynamic shape input data. Tagging @bixia1 for review.

### CODE REVIEW

**Comment by gbaned** [comment]
@tfeher Can you please resolve conflicts? Thanks!

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc [hunk 1]
```diff
@@ -4523,40 +4525,61 @@ Status ConvertPack(OpConverterParams* params) {
...
   }
 
-  // Convert axis from the TensorFlow format to TensorRT format.
-  const nvinfer1::Dims dims = inputs.at(0).GetTrtDims();
+  // Convert axis from the TensorFlow format to TensorRT format. First we check
+  // if any of the inputs is a weight. If yes then we use its shape, because
+  // it is conversion time constant, otherwise we just select the first input.
+  int idx = 0;
+  for (int i = 0; i < inputs.size(); i++) {
+    if (inputs.at(i).is_weights()) {
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:4555]: Can you change this to find an input with all dimension values known instead? It is possible that there are multiple non-weights inputs, some with dynamic dimension values but others with static dimension values, right?

Can you also add a test case for selecting a tensor with static shape?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 2]
```diff
@@ -5681,152 +5669,149 @@ NodeDef GetPackNodeDef(DataType dtype, int num_inputs, int axis) {
...
+          /*input_shapes=*/{{1, 2, 3}, {1, 2, 3}},
+          /*input_values=*/common_input,
+          /*axis=*/-5,
+          /*expected_output_dims=*/{},
+          /*expected_output=*/{},
+          Status{error::INVALID_ARGUMENT,
+                 "Axis value of -5 is out of bounds, must be in"
+                 " range [-4, 4), at my_pack"},
+      },
+      // Axis is batch dimension, should fail.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:5715]: Fail only in implicit batch mode?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc [hunk 3]
```diff
@@ -5681,152 +5669,149 @@ NodeDef GetPackNodeDef(DataType dtype, int num_inputs, int axis) {
...
     std::vector<std::vector<int>> input_shapes;
-    std::vector<std::vector<CType>> input_values;
+    std::vector<std::vector<float>> input_values;
     int axis;
     std::vector<int> expected_output_dims;
-    std::vector<CType> expected_output;
+    std::vector<float> expected_output;
+    Status conversion_status;
+    Status runtime_status;
+    bool weight_input;
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:5685]: second_input_is_weight or input_1_is_weight?

**Comment by gbaned** [comment]
@tfeher Can you please check @bixia1's comments and keep us posted. Thanks!

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc (hunk 1 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: That is a good point. Yes, it is possible to have multiple non-weight inputs where some could have static shape. I have changed the condition and added a test (the last test).

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 3 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: changed to input_1_is_weight

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc (hunk 2 above)
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: That is right, corrected.

**Review by tfeher — COMMENTED** [review]
Thank you @bixia1 for the review, I have addressed the issues.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**